### ISSUE TITLE
Convert yolov5s to tensorrt with hyp.yaml

### ISSUE THREAD

**Issue by MiladGhorbaniG** [wang-xinyu/tensorrtx#674]
## Env

- Jetson Nano
- Ubuntu18.04, 
- Cuda = 10.2
- TRT = TensorRT 7.1.3

## About this repo

- model yolov5

## Your problem

Hi @wwqgtxx ,
I trained a yolov5s6 with the following hyperparameters.

          lr0: 0.0113
          lrf: 0.225
          momentum: 0.814
          weight_decay: 0.00074
          warmup_epochs: 2.3
          warmup_momentum: 0.761
          warmup_bias_lr: 0.0624
          box: 0.035
          cls: 0.717
          cls_pw: 0.912
          obj: 1.24
          obj_pw: 1.54
          iou_t: 0.2
          anchor_t: 4.37
          anchors: 2.0
          fl_gamma: 0.0
          hsv_h: 0.0174
          hsv_s: 0.814
          hsv_v: 0.19
          degrees: 0.0
          translate: 0.0871
          scale: 0.826
          shear: 0.0
          perspective: 0.0
          flipud: 0.0
          fliplr: 0.5
          mosaic: 0.872
          mixup: 0.0

Then converted to .wts using: 
 `python gen_wts.py yolov5s.pt`
 But I couldn't convert to .engine when I used:
 `sudo ./yolov5 -s yolov5s.wts yolov5s.engine s`

How can I change config files to convert the model to TRT correctly?

**Comment by wang-xinyu** [comment]
for yolov5s6:

```
sudo ./yolov5 -s yolov5s6.wts yolov5s6.engine s6
```

**Comment by MiladGhorbaniG** [comment]
I used 
`sudo ./yolov5 -s yolov5s6.wts yolov5s6.engine s6`
but it outputs the following error
    Building engine, please wait for a while...
    [08/16/2021-12:56:10] [E] [TRT] (Unnamed Layer* 345) [Convolution]: kernel weights has count 1536 but 2304 was expected
    [08/16/2021-12:56:10] [E] [TRT] (Unnamed Layer* 345) [Convolution]: count of 1536 weights in kernel, but kernel dimensions (1,1) with 128 input channels, 18 output channels and 1 groups were specified. Expected Weights count is 128 * 1*1 * 18 / 1 = 2304
    [08/16/2021-12:56:10] [E] [TRT] Could not compute dimensions for (Unnamed Layer* 345) [Convolution]_output, because the network is not valid.
    [08/16/2021-12:56:10] [E] [TRT] Network validation failed.
    Build engine successfully!
    yolov5: /home/jetson/wang-xinyu/new/tensorrtx/yolov5/yolov5.cpp:245: void APIToModel(unsigned int, nvinfer1::IHostMemory**, bool&, float&, float&, std::__cxx11::string&): Assertion `engine != nullptr' failed.
    Aborted

**Comment by wang-xinyu** [comment]
Try the official model first. And don't forget to change CLASS_NUM.

**Comment by MiladGhorbaniG** [comment]
Prior to using custom parameters, converting the model to TRT was successful. However, the model with customized hyperparameters is more accurate. Is there a way to convert it?

**Comment by wang-xinyu** [comment]
If hyps did not affect the model arch, nothing needs to be changed.

**Comment by MiladGhorbaniG** [comment]
I guess that these hyperparameters might affect the model architecture, 

anchor_t: 4.0  # anchor-multiple threshold
anchors: 3  # anchors per output layer (0 to ignore)

Can the anchors configurations be changed?

**Comment by wang-xinyu** [comment]
Have you adapted the class_num?

**Comment by wang-xinyu** [comment]
The error message shows `convolution`, not related to anchor.

**Comment by MiladGhorbaniG** [comment]
I used the default parameter for `anchor_t:4.0` and commented `anchors:3` and then I could successfully convert to TRT. So it seems that these 2 parameters caused the issue.

**Comment by wang-xinyu** [comment]
Will this param affect number of conv layers in the model?

**Comment by wang-xinyu** [comment]
Can you check if models.yolo.Detect contains conv layer?

**Comment by MiladGhorbaniG** [comment]
Yes it has some convolutions. `self.m` are the convolutions in Detect layer as in the definition:

`
class Detect(nn.Module):
    stride = None  # strides computed during build
    onnx_dynamic = False  # ONNX export parameter

    def __init__(self, nc=80, anchors=(), ch=(), inplace=True):  # detection layer
        super().__init__()
        self.nc = nc  # number of classes
        self.no = nc + 5  # number of outputs per anchor
        self.nl = len(anchors)  # number of detection layers
        self.na = len(anchors[0]) // 2  # number of anchors
        self.grid = [torch.zeros(1)] * self.nl  # init grid
        a = torch.tensor(anchors).float().view(self.nl, -1, 2)
        self.register_buffer('anchors', a)  # shape(nl,na,2)
        self.register_buffer('anchor_grid', a.clone().view(self.nl, 1, -1, 1, 1, 2))  # shape(nl,1,na,1,1,2)
        self.m = nn.ModuleList(nn.Conv2d(x, self.no * self.na, 1) for x in ch)  # output conv
        self.inplace = inplace  # use in-place ops (e.g. slice assignment)

    def forward(self, x):
        # x = x.copy()  # for profiling
        z = []  # inference output
        for i in range(self.nl):
            x[i] = self.m[i](x[i])  # conv
            bs, _, ny, nx = x[i].shape  # x(bs,255,20,20) to x(bs,3,20,20,85)
            x[i] = x[i].view(bs, self.na, self.no, ny, nx).permute(0, 1, 3, 4, 2).contiguous()

            if not self.training:  # inference
                if self.grid[i].shape[2:4] != x[i].shape[2:4] or self.onnx_dynamic:
                    self.grid[i] = self._make_grid(nx, ny).to(x[i].device)

                y = x[i].sigmoid()
                if self.inplace:
                    y[..., 0:2] = (y[..., 0:2] * 2. - 0.5 + self.grid[i]) * self.stride[i]  # xy
                    y[..., 2:4] = (y[..., 2:4] * 2) ** 2 * self.anchor_grid[i]  # wh
                else:  # for YOLOv5 on AWS Inferentia https://github.com/ultralytics/yolov5/pull/2953
                    xy = (y[..., 0:2] * 2. - 0.5 + self.grid[i]) * self.stride[i]  # xy
                    wh = (y[..., 2:4] * 2) ** 2 * self.anchor_grid[i].view(1, self.na, 1, 1, 2)  # wh
                    y = torch.cat((xy, wh, y[..., 4:]), -1)
                z.append(y.view(bs, -1, self.no))

        return x if self.training else (torch.cat(z, 1), x)
`

**Comment by huyhoang17** [comment]
@wang-xinyu 
Hi Wang, 
I met the same issue with the custom model, I change both the number of filters & `depth_multiple/width_multiple` param in yaml file for building my custom model. The command I used
```bash
./yolov5 -s custom_model.wts custom_model.engine s
```

![image](https://user-images.githubusercontent.com/15128654/130061012-19c42c0f-d8c8-4f18-8b91-54c9d3cff2ec.png)

I also change the number of classes in `yololayer.h` file and re-build
I changed the model parameter by half, how can I fix this issue? Thank you!

**Comment by wang-xinyu** [comment]
@huyhoang17 
```
// For example Custom model with depth_multiple=0.17, width_multiple=0.25 in yolov5.yaml
sudo ./yolov5 -s yolov5_custom.wts yolov5.engine c 0.17 0.25
```

**Comment by wang-xinyu** [comment]
@MiladGhorbaniG So you need to adapt the conv related params accordingly in .cpp

**Comment by aHodor-z** [comment]
@huyhoang17 
Hello, I have the same problem as you, my depth_multiple/width_multiple are 0.33 and 0.5 respectively, and the number of categories is 3.

But execute `sudo ./yolov5 -s yolov5_custom.wts yolov5.engine c 0.33 0.5`

I also change the number of classes in yololayer.h file and re-build

How did you solve this problem? Thank you!
![374B8BD9AC6089ECC3349E589DE2377A](https://user-images.githubusercontent.com/71739449/132273156-874bd789-e19f-479d-8df1-31fe1c00fe2b.jpg)

**Comment by huyhoang17** [comment]
@aHodor-z 
I used this package and converted model to Trt format without error, you can try it: https://github.com/Media-Smart/volksdep
You need to convert the yolov5 model to ONNX format first.

**Comment by ShanQiong** [comment]
> I met this issue tonight，you need to add static const int Anchor_NUM = Yolo::CHECK_COUNT, and then change the code on line 84, 89, 94, 3 to Anchor_NUM, then it works.

Yes, I forgot to reply, I also doing in this way, Thanks

**Comment by sicarioakki** [comment]
> I met this issue tonight，you need to add static const int Anchor_NUM = Yolo::CHECK_COUNT, and then change the code on line 84, 89, 94, 3 to Anchor_NUM, then it works.

Can you please explain this with a little more description? I cannot seem to understand your solution and I face the same issue.
